### DISCUSSION TITLE
Current implementation doesn't support at least one key scenario

### DISCUSSION THREAD

**Discussion by donniehale-awh** [dotnet/aspnetcore#24680] [General]
The canonical means of registering an EF Core DbContext with the service collection is:

`services.AddDbContext<MyDbContext>(o => o.UseSqlServer(configuration.GetConnectionString("MyConnectionString")));`

As far as I can tell that isn't possible in the current IHostingStartup approach because the actual IConfiguration isn't available at the time that `Configure(IWebHostBuilder builder)` is called. When the EF Core portion of an application is (properly, IMO) implemented as its own class library, there needs to be some means of initializing the DbContext and registering it with DI.

Am I missing something (very possible)? Or is there some other supported means of doing this at application startup without the asp.net core application taking a dependency on the EF Core class library (something that we want to avoid at pretty much any cost).

Thanks.

**Comment by donniehale-awh** [comment]
In case it isn't clear, I'm referring to:

https://github.com/dotnet/aspnetcore/search?q=IHostingStartup&unscoped_q=IHostingStartup

and this documentation:

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-3.1

(I started the discussion from the IHostingStartup page and thought it would be specifically about that. My apologies.)

**Reply by davidfowl** [reply]
It would look like this:

```C#
.ConfigureServices((context, services) =>
{
    services.AddDbContext<MyDbContext>(o => o.UseSqlServer(context.Configuration.GetConnectionString("MyConnectionString")));
})
```

**Reply by donniehale-awh** [reply]
I did a lousy job explaining the problem I'm trying to solve. My apologies.

I need to do the equivalent of `ConfigureServices` in an assembly other than the startup assembly. This article documents a level of supported extensibility: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-3.1. Per those docs, I add a call to UseSetting in Program.cs:

`
        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseSetting(WebHostDefaults.HostingStartupAssembliesKey, "MyAssembly");
                    webBuilder.UseStartup<ApiStartup>();
                });
`

Doing that causes my assembly to get loaded. Then I place a `HostingStartup` assembly attribute in my assembly:

`[assembly: HostingStartup(typeof(MyAssembly.DataStartup))]`

And I have the `DataStartup` class implement `IHostingStartup`:

`public class DataStartup : IHostingStartup`. This includes implementing the interface's `Configure` method which is passed an `IWebHostBuilder`. Doing that I can use `builder.ConfigureServices` to do normal DI configuration.

What that does not seem to allow me to do is make the `AddDbContext` call where the connection string comes from configuration. That's because when the `IHostingStartup.Configure(IWebHostBuilder builder)` method is called, the configuration (`IConfiguration`) hasn't been "built" yet and made available for looking up configuration values. The startup assembly's Startup.cs class is instantiated after the configuration has been built is supplied that `IConfiguration`. But in this `IHostingStartup` approach, the configuration does not seem to be available.

I'd be happy to be corrected or shown what I'm missing. Thank you again.
